src/raw/mod.rs

```diff
@@ -0,0 +1,1634 @@
+            }
+            ElementType::Undefined => 0,
+            ElementType::ObjectId => 12,
+            ElementType::Boolean => 1,
+            ElementType::DateTime => 8,
+            ElementType::Null => 0,
+            ElementType::RegularExpression => {
+                let regex = match read_nullterminated(&self.doc.data[valueoffset..]) {
+                    Ok(regex) => regex,
+                    Err(err) => return Some(Err(err)),
+                };
+
+                let options =
+                    match read_nullterminated(&self.doc.data[valueoffset + regex.len() + 1..]) {
+                        Ok(options) => options,
+                        Err(err) => return Some(Err(err)),
+                    };
+
+                regex.len() + options.len() + 2
+            }
+            ElementType::DbPointer => {
+                let string_size =
+                    4 + i32_from_slice(&self.doc.data[valueoffset..valueoffset + 4]) as usize;
+
+                let id_size = 12;
+
+                if self.doc.data[valueoffset + string_size - 1] != 0 {
+                    return Some(Err(RawError::MalformedValue(
+                        "DBPointer string not null-terminated".into(),
+                    )));
+                }
+
+                string_size + id_size
+            }
+            ElementType::JavaScriptCode => {
+                let size =
+                    4 + i32_from_slice(&self.doc.data[valueoffset..valueoffset + 4]) as usize;
+
+                if self.doc.data[valueoffset + size - 1] != 0 {
+                    return Some(Err(RawError::MalformedValue(
+                        "javascript code not null-terminated".into(),
+                    )));
+                }
+
+                size
+            }
+            ElementType::Symbol => {
+                4 + i32_from_slice(&self.doc.data[valueoffset..valueoffset + 4]) as usize
+            }
+            ElementType::JavaScriptCodeWithScope => {
+                let size = i32_from_slice(&self.doc.data[valueoffset..valueoffset + 4]) as usize;
+
+                if self.doc.data[valueoffset + size - 1] != 0 {
+                    return Some(Err(RawError::MalformedValue(
+                        "javascript with scope not null-terminated".into(),
+                    )));
+                }
+
+                size
+            }
+            ElementType::Int32 => 4,
+            ElementType::Timestamp => 8,
+            ElementType::Int64 => 8,
+            ElementType::Decimal128 => 16,
+            ElementType::MaxKey => 0,
+            ElementType::MinKey => 0,
+        };
+
+        let nextoffset = valueoffset + element_size;
+        self.offset = nextoffset;
+
+        Some(Ok((
+            key,
+            elem::RawBson::new(element_type, &self.doc.data[valueoffset..nextoffset]),
+        )))
+    }
+}
+
+/// A BSON array referencing raw bytes stored elsewhere.
+pub struct RawArray {
```

Comment:
can we break `RawArray` and `RawDocument` + `RawDocumentRef` into their own files in the `raw` module?

Reply:
I've split what previously existing in the top-level `raw` module into the `doc`, `array`, and `test` (private) submodules, leaving behind only re-exports of the public items from those submodules and a few helper functions to be used by the submodules.